### PR TITLE
Forecasting Interface

### PR DESCRIPTION
**Pull request by damonbayer**
Closes https://github.com/CDCgov/multisignal-epi-inference/issues/169
Partially addresses https://github.com/CDCgov/multisignal-epi-inference/issues/249.

This PR

1. Upgrades minimum version of`jax` and `numpyro` in light of https://github.com/pyro-ppl/numpyro/issues/1827.
2. Modifies models to have a clearer definition of some time-related concepts.
- `n_timepoints_to_simulate` is always the length of the "data" (whether or not it is observed). I would like to rename it to `n_datapoints` in a separate PR.
- Observations begin after the seeding period and the padding period. Thus the total length of the latent infections is `n_timepoints_to_simulate` + `seeding_period_length` + `padding`
3. Cleans up `hospital_admissions_model.qmd` in light of (2) and to reduce code repetition.
4. Simplifies `admissionsmodel.py` and `rtinfectionsrenewalmodel.py` to avoid unnecessary `if-else` paths.
5. Modifies `simplerandomwalk.py` to use `numpyro.contrib.control_flow.scan` rather than an `expand`ed distribution, which allows for forecasting to "just work."
6. Adds a forecasting example to `hospital_admissions_model.qmd`.

# Original PR text:

Currently working on refactoring the `SimpleRandomWalkProcess` using `scan`, which should make forecasting possible. It works in forward simulation, but is broken in inference.

This works:
```python
from pyrenew.process import SimpleRandomWalkProcess
from numpyro import distributions as dist
import numpyro
srwp = SimpleRandomWalkProcess(dist.Normal(0, 1))
with numpyro.handlers.seed(rng_seed=0):
    x = srwp.sample(n_timepoints=10, init=0.0)
x
```

But `pyrenew_demo.qmd` fails during inference.

It says:
```pycon
TypeError: body_fun output and input must have identical types, got
('ShapedArray(int32[], weak_type=True)', ['ShapedArray(float32[33])', 'DIFFERENT ShapedArray(int32[], weak_type=True) vs. ShapedArray(float0[])', 'ShapedArray(float32[])'], []).
```

### CODE REVIEW

**Comment by damonbayer** [comment]
@dylanhmorris @gvegayon I could use your eyes on this.

**Comment by damonbayer** [comment]
Note, I get the same error when trying to perform inference on the [example given in the `scan` documentation](https://num.pyro.ai/en/stable/primitives.html#scan):

```python
import numpy as np
import numpyro
import numpyro.distributions as dist
from numpyro.contrib.control_flow import scan

def gaussian_hmm(y=None, T=10):
    def transition(x_prev, y_curr):
        x_curr = numpyro.sample('x', dist.Normal(x_prev, 1))
        y_curr = numpyro.sample('y', dist.Normal(x_curr, 1), obs=y_curr)
        return x_curr, (x_curr, y_curr)

    x0 = numpyro.sample('x_0', dist.Normal(0, 1))
    _, (x, y) = scan(transition, x0, y, length=T)
    return (x, y)


with numpyro.handlers.seed(rng_seed=0):  # generative
    x, y = gaussian_hmm()

from jax import random
from numpyro.infer import MCMC, NUTS
nuts_kernel = NUTS(gaussian_hmm)

mcmc = MCMC(nuts_kernel, num_warmup=500, num_samples=1000)
rng_key = random.PRNGKey(0)

mcmc.run(rng_key, y=y)
```

**Comment by dylanhmorris** [comment]
> Note, I get the same error when trying to perform inference on the [example given in the scan documentation](https://num.pyro.ai/en/stable/primitives.html#scan):

I am able to run your code block without a problem.

**Comment by damonbayer** [comment]
@dylanhmorris and I have determined that this works with jax 0.4.28, but not jax 0.4.30

**Comment by damonbayer** [comment]
This is now working with downgraded jax. Will continue on the PR.

**Comment by damonbayer** [comment]
Forecasting "just works" with this model change. You can do model.posterior_predictive(n_timepoints_to_simulate = an arbitrary number) to forecast arbitrarily far into the future.

image


Keeping this as a draft for now because changes to tutorials are about to land (https://github.com/CDCgov/multisignal-epi-inference/pull/234 & https://github.com/CDCgov/multisignal-epi-inference/pull/191). Concerns raised in https://github.com/CDCgov/multisignal-epi-inference/issues/249 will partially be addressed here.

**Comment by damonbayer** [comment]
@dylanhmorris @gvegayon This is ready for review, but I'm still thinking about a better way to handle `padding` throughout.

**Comment by damonbayer** [comment]
The most recent commits adopt a clearer definition of some time-related concepts.

- `n_timepoints_to_simulate` is always the size of the thing that is observed (I would like to rename it to `n_datapoints` in a separate PR.
- The observations begin after the seeding period and the padding period. Thus the total length of the latent infections is `n_timepoints_to_simulate` + `seeding_period_length` + `padding`

**Comment by damonbayer** [comment]
@dylanhmorris @gvegayon This is ready for review. I have added a summary of all the changes in the at the top of this page.

**Comment by AFg6K7h4fhy2** [comment]
> @dylanhmorris @gvegayon This is ready for review. I have added a summary of all the changes in the at the top of this page.

Would appreciate this as it will alter some aspects of #162

File: model/src/pyrenew/model/admissionsmodel.py [hunk 1]
```diff
@@ -199,35 +201,17 @@ def sample(
             latent_infections=basic_model.latent_infections,
             **kwargs,
         )
-        i0_size = len(latent_hosp_admissions) - n_timepoints
         if self.hosp_admission_obs_process_rv is None:
             observed_hosp_admissions = None
```
Comment (damonbayer) [inline model/src/pyrenew/model/admissionsmodel.py:205]: I think these can be removed

File: model/src/test/test_random_walk.py [hunk 2]
```diff
@@ -35,9 +35,9 @@ def test_rw_samples_correctly_distributed():
         [0, 2.253, -3.2521, 1052, 1e-6], [1, 0.025, 3, 1, 0.02]
     ):
         rw_normal = SimpleRandomWalkProcess(dist.Normal(step_mean, step_sd))
-        init_arr = jnp.array([532.0])
+        init = 532.0
         with numpyro.handlers.seed(rng_seed=62):
-            samples, *_ = rw_normal.sample(n_samples, init=init_arr)
+            samples, *_ = rw_normal.sample(n_samples, init=init)
```
Comment (AFg6K7h4fhy2) [inline model/src/test/test_random_walk.py:40]: Minor comment, but historically speaking, I've leaned against using `init` alone as a variable name given its proximity to `__init__`; I am not sure if this is outlined in any [PEP](https://peps.python.org/), but [PEP 8](https://peps.python.org/pep-0008/) does not explicitly forbid this practice. I feel as though there is something _slightly_ more sensible that can be done, but I am fine with `init` being used and thought my mental reaction was worth recording.

Comment (damonbayer) [reply model/src/test/test_random_walk.py]: addressed in https://github.com/CDCgov/multisignal-epi-inference/pull/241/commits/bddc731ac171e679065e0ca01d9f63f1bb487c9e

File: docs/source/tutorials/basic_renewal_model.qmd [hunk 3]
```diff
@@ -214,7 +214,9 @@ model1.run(
     num_samples=1000,
     data_observed_infections=sim_data.observed_infections,
     rng_key=jax.random.PRNGKey(54),
-    mcmc_args=dict(progress_bar=False, num_chains=2, chain_method="sequential"),
+    mcmc_args=dict(
```
Comment (dylanhmorris) [inline docs/source/tutorials/basic_renewal_model.qmd:217]: Not for this PR but these diffs remind me that we need to get something working for shared formatting of `.qmd` files

File: model/src/pyrenew/process/simplerandomwalk.py [hunk 4]
```diff
@@ -62,12 +63,20 @@ def sample(
 
         if init is None:
             init = npro.sample(name + "_init", self.error_distribution)
-        diffs = npro.sample(
-            name + "_diffs",
-            self.error_distribution.expand((n_timepoints - 1,)),
+
+        def transition(x_prev, _):
+            # numpydoc ignore=GL08
+            diff = npro.sample(name + "_diffs", self.error_distribution)
```
Comment (dylanhmorris) [inline model/src/pyrenew/process/simplerandomwalk.py:69]: What gets tracked by this `sample` call when inference is performed? Is only the final value in the scan logged? Is a vector of length `n_timepoints - 1` created?

File: docs/source/tutorials/hospital_admissions_model.qmd [hunk 5]
```diff
@@ -535,21 +540,29 @@ And plotting the results:
...
+    obs_signal=np.pad(
+        daily_hosp_admits.astype(float),
+        (gen_int_array.size + pad_size, 0),
+        constant_values=np.nan,
+    ),
 )
 ```
 
 We will use ArviZ to visualize the posterior/prior predictive distributions.
+By increasing `n_timepoints_to_simulate`, we can perform forecasting.
```
Comment (dylanhmorris) [inline docs/source/tutorials/hospital_admissions_model.qmd:552]: Potentially a bit confusing to have this line here but then go into prior predictive analysis and come back to forecasting further down, without mentioning that that's what it is. Perhaps insert something highlighting forecasting here?

https://github.com/CDCgov/multisignal-epi-inference/blob/1795426650682eab3ab7367e20ea8a5365a5d5e4/docs/source/tutorials/hospital_admissions_model.qmd#L602

**Review by dylanhmorris — CHANGES_REQUESTED** [review]
Looks good, thanks @damonbayer. A couple questions and one major request.

Request: can we have a forecasting unit test that confirms stochastic processes are projected forward as we would expect (e.g. that forecast-period steps in a random walk are generated conditional on the posterior sampled value of the step s.d.)

File: docs/source/tutorials/basic_renewal_model.qmd (hunk 3 above)
Comment (damonbayer) [reply docs/source/tutorials/basic_renewal_model.qmd:217]: I have posted on Teams and updated the this issue in this repo https://github.com/CDCgov/multisignal-epi-inference/issues/159

File: model/src/pyrenew/process/simplerandomwalk.py (hunk 4 above)
Comment (damonbayer) [reply model/src/pyrenew/process/simplerandomwalk.py:69]: `Rt_transformed_rw_diffs[0]`, `Rt_transformed_rw_diffs[1]`, … `Rt_transformed_rw_diffs[n]` are all created.

Comment (dylanhmorris) [reply model/src/pyrenew/process/simplerandomwalk.py:69]: Excellent.

File: docs/source/tutorials/basic_renewal_model.qmd (hunk 3 above)
Comment (dylanhmorris) [reply docs/source/tutorials/basic_renewal_model.qmd:217]: Thanks!

File: docs/source/tutorials/hospital_admissions_model.qmd (hunk 5 above)
Comment (damonbayer) [reply docs/source/tutorials/hospital_admissions_model.qmd]: addressed in https://github.com/CDCgov/multisignal-epi-inference/pull/241/commits/a84ca20e4e1183033df3c65e0c3b061ad2f410e7

File: docs/source/tutorials/hospital_admissions_model.qmd [hunk 6]
```diff
@@ -535,21 +540,29 @@ And plotting the results:
...
+        daily_hosp_admits.astype(float),
+        (gen_int_array.size + pad_size, 0),
+        constant_values=np.nan,
+    ),
 )
 ```
 
-We will use ArviZ to visualize the posterior/prior predictive distributions.
+We will use ArviZ to visualize the posterior and prior predictive distributions.
+By increasing `n_timepoints_to_simulate`, we can perform forecasting in the posterior predictive.
```
Comment (dylanhmorris) [inline docs/source/tutorials/hospital_admissions_model.qmd:552]:
```suggestion
By increasing `n_timepoints_to_simulate`, we can perform forecasting using the posterior predictive distribution.
```

**Comment by damonbayer** [comment]
> Looks good, thanks @damonbayer. A couple questions and one major request.
> 
> Request: can we have a forecasting unit test that confirms stochastic processes are projected forward as we would expect (e.g. that forecast-period steps in a random walk are generated conditional on the posterior sampled value of the step s.d.)

We do not currently support posterior updates of the step s.d., as far as I can tell.

File: docs/pyproject.toml [hunk 7]
```diff
@@ -9,8 +9,8 @@ package-mode = false
 [tool.poetry.dependencies]
 python = "^3.12"
 sphinx = "^7.2.6"
-jax = "^0.4.25"
-jaxlib = "^0.4.25"
+jax = "0.4.29"
+jaxlib = "^0.4.29"
```
Comment (damonbayer) [inline docs/pyproject.toml:13]: Numpyro 0.15.1 is released and makes this unnecessary https://github.com/pyro-ppl/numpyro/releases/tag/0.15.1

**Review by dylanhmorris — APPROVED** [review]
LGTM, thanks @damonbayer!

**Pull request merged**